[1단계 - 자동차 경주 구현] 알렉스(이영환) 미션 제출합니다.

yxxnghwan opened the pull request:
잘 부탁드립니다 선배님!😁

한 가지 궁금한 게 있습니다.

CarNameValidator.validate() 함수에서 첫 번째 조건문에 조건식이 name.isBlank()인데요,
다른 조건식들은 이 조건이 어떤 조건식인지 함수로 분리해서 네이밍을 통해 의도를 드러냈는데,
저 조건식은 딱 한 가지 조건만 체크하고 그 조건은 Java의 기본 String API가 제공하는 메소드로 해결이 됩니다.
이런 경우에도 다른 조건들과 통일성을 위해 딱 한 가지 조건이라도 함수로 빼는 것이 좋을까요? 아니면 이미 isBlank()라는 메서드가 의도를 드러내고 있으니 현 상태로 두는게 좋을까요?

yxxnghwan committed:
docs(calculator.README): 구현할 기능 목록 작성

yxxnghwan committed:
test(calculator): StringCalculator 기본 단위 테스트 작성

yxxnghwan committed:
feat(calculator): 0 또는 숫자만 반환하는 경우 작성

yxxnghwan committed:
feat(calculator): 커스텀 구분자가 있는 경우 구분자 추가, 파라미터 추출

yxxnghwan committed:
feat(calculator): 양의 정수가 아닌 경우 예외처리 작성

yxxnghwan committed:
feat(calculator): 숫자 합산

yxxnghwan committed:
fix(calculator): null값이 들어올 경우 NullPointerException 발생하던 버그 수정

yxxnghwan committed:
fix(calculator): 양의 정수 체크 조건을 반대로 적용시킨 버그 수정

yxxnghwan committed:
feat(calculator): main class, UI 구현

yxxnghwan committed:
refactor(calculator): 메소드 분리로 indent 및 라인 수 조절

yxxnghwan committed:
docs(racing-car): 구현할 기능 목록 작성

yxxnghwan committed:
test(racing-car): 난수 생성함수 범위 테스트 작성

yxxnghwan committed:
chore(racing-car): utils 패키지로 이동

yxxnghwan committed:
feat(racing-car): RandomIntegerGenerator 구현

yxxnghwan committed:
docs(racing-car): Exception 종류 수정

yxxnghwan committed:
test(racing-car): CarNameValidatorTest 작성

yxxnghwan committed:
feat(racing-car): CarNameValidator 구현

yxxnghwan committed:
test(racing-car): RoundNumberValidatorTest 작성

yxxnghwan committed:
fix(racing-car): RoundNumberValidatorTest 파라미터 타입 수정

yxxnghwan committed:
feat(racing-car): RoundNumberValidator 구현

yxxnghwan committed:
test(racing-car): CarTest 작성

yxxnghwan committed:
test(racing-car): RacingServiceTest 작성

yxxnghwan committed:
feat(racing-car): Car 초기 구현

yxxnghwan committed:
test(racing-car): CarTest 전진 여부 테스트 추가 검증

랜덤 숫자가 4보다 작을 경우 전진하지 않는지 검사

yxxnghwan committed:
feat(racing-car): RacingService 구현

service 구현에 따른 추가적인 Car.isWinner 메소드 작성

yxxnghwan committed:
feat(racing-car): CarsGenerator 구현

사용자 입력값(String) -> List<Car> 형태로 변환 시키기 위 해 정의

yxxnghwan committed:
test(racing-car): CarsGeneratorTest 작성

yxxnghwan committed:
feat(racing-car): UI 클래스 구현

yxxnghwan committed:
refactor(racing-car): 적절한 파라미터명으로 수정

names -> userInput 변경

yxxnghwan committed:
feat(racing-car): 잘못된 값 입력 시 재입력받도록 수정

yxxnghwan committed:
feat(racing-car): Main 로직 구현

yxxnghwan committed:
refactor(calculator): 리팩토링 및 상수화 작업

yxxnghwan committed:
refactor(calculator): 정규표현식 네이밍 명확하게 수정

yxxnghwan committed:
refactor(racing-car): 리팩토링 및 상수화 작업

yxxnghwan committed:
refactor(calculator): CalculatorInput 클래스 상수화 작업

yxxnghwan committed:
refactor(calculator): StringCalculator 메서드 분리

커스텀 구분자를 가지는 입력인지 판단하는 조건문을 메서드로 분리

sonypark commented on docs/README-racing-car.md:
```diff
@@ -0,0 +1,14 @@
+ - 회차마다 자동차의 위치를 `-`로 출력해준다.
+ - 경주가 완료되면 자동차의 위치를 기반으로 우승자를 출력한다. 
+ - 우승자는 여러 명일 수 있다.
+ - 
```
README 잘 작성해주셨네요! 👍
README에 정의된 기능을 구현할 때마다 체크할 수 있도록 [체크박스 형태](https://www.w3schools.io/file/markdown-checkbox-github/)로 작성해도 좋을 것 같아요~

yxxnghwan replied:
피드백 반영하여 체크박스 형태로 수정하고, 목록 요소 하나당 기능이 한 가지만 적혀 있는 것이 좋을 것 같아서 더 잘게 나누었습니다! 피드백 감사합니다 소니😄

sonypark commented on src/main/java/calculator/CalculatorInput.java:
```diff
@@ -0,0 +1,16 @@
+		System.out.print(INPUT_MESSAGE);
+		String stringInput = CalculatorMain.sc.nextLine();
+		if (stringInput.contains(CUSTOM_SEPARATOR_START_SIGNATURE )) {
+			return stringInput + CUSTOM_SEPARATOR_END_SIGNATURE + CalculatorMain.sc.nextLine();
```
이 부분 로직이 의도한대로 동작하는게 맞나요?
Custom delimiter 입력에 대한 로직인 것 같은데
어플리케이션 실행 후 `//;\n1;2;3` 를 입력하면 어떻게 될까요?

yxxnghwan replied:
어플리케이션 테스트를 소홀히 했나봐요... 
첫 PR에서는 인텔리 제이에서 테스트코드에 있는 문자열을 복사 붙여넣기로 콘솔창에 입력했더니
```
//;
1;2;3
```
이렇게 나뉘어져서 들어와서 저렇게 구현했었어요,

직접 입력하거나, 브라우저에 있는 `//;\n1;2;3`를 입력했을 때는
이스케이프 문자를 인식하지 않고 문자 그대로 들어오더라구요.
`"//;\\n1;2;3"` 이렇게 선언한 문자열과 equals 했을 때 true를 반환하는 것을 확인했어요.

해당 부분을 확인하고 거기에 맞게 수정했습니다.

sonypark commented on src/main/java/calculator/CalculatorMain.java:
```diff
@@ -0,0 +1,14 @@
+		int result = StringCalculator.calculate(calculatorInput.userInput());
+		CalculatorOutput calculatorOutput = new CalculatorOutput(result);
+		calculatorOutput.OutputResult();
+	}
```
👍

sonypark commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,72 @@
+	private static final String BASIC_SEPARATOR_COMMA = ",";
+	private static final String BASIC_SEPARATOR_COLON = ":";
+	private static final String ZERO_AND_POSITIVE_INTEGER_REGEX = "^(0|[1-9][0-9]*)$";
+	private static final String INVALID_POSITIVE_INTEGER_ERROR_MESSAGE = "양의 정수를 입력 해주세요.";
```
상수 추출 👍

sonypark commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,72 @@
+		checkRightPositiveInteger(numberStringList);
+		List<Integer> numberList = toNumberList(numberStringList);
+		return numberList.stream().mapToInt(num -> num).sum();
+	}
```
메서드 분리 👍

sonypark commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,72 @@
+		return numberStringList.stream()
+			.mapToInt(Integer::parseInt)
+			.boxed()
+			.collect(Collectors.toList());
```
이렇게 할 수도 있겠네요 :)

```suggestion
		return numberStringList.stream()
		         .map(Integer::valueOf)
			.collect(Collectors.toList());
```

yxxnghwan replied:
박싱, 언박싱과 `parseInt()`, `valueOf()`의 차이점에 대해 짚어볼 수 있었어요! 감사해요 소니😁

sonypark commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,48 @@
+	public int hashCode() {
+		return Objects.hash(getName(), getPosition());
+	}
+}
```
이 메서드를 오버라이드 해준 이유가 있을까요?

yxxnghwan replied:
CarsGeneratorTest에서 입력된 문자열로 자동차 객체를 제대로 생성하는 지 확인하기 위해 구현했는데요,
컬렉션을 테스트하는 `contains()`메서드가 `equals()` 기준으로 동작하더라구요.

프로덕션 코드에서는 `equals()`를 전혀 사용하지 않고 테스트 코드를 위해 오버라이드했습니다.
만약 이 경우에 오버라이드 하는 것이 적절하지 않다면, 테스트 코드를 다른 방식으로 바꾸는 것이 좋을까요?

지금 생각나는 대안은 Car 클래스를 상속받는 테스트 전용 클래스를 만들어서 `equals()`를 오버라이드 하고 테스트에서만 사용하는 방법이 떠오르네요.

의견 부탁드려요 소니!

yxxnghwan replied:
금요일에 우테코 TDD 강의가 있었는데요, 해당 부분을 질문했는데 코치분들 의견이 갈렸어요,,, 소니 의견은 어떤지 여쭤볼 수 있을까요?

Q. 프로덕션 코드에서 equals를 사용하지 않는데, 테스트를 위해서 equals를 오버라이드 해도 될까요?

네오: 된다. 우리는 객체지향 코드를 만드는 사람이기 때문에 만드는 객체들은 객체의 기본 원칙인 equals 와 hashCode 를 정의해 줘야 한다. 꼭 equals를 사용 하지 않더라도..

구구: equals를 테스트 때문에 오버라이드 하는 건 제 경우에는 필요할때만 할 것 같아요.

sonypark replied:
네 좋은 고민인 것 같습니다. 우선 테스트를 왜 하는가에 대해서 생각해 보면 좋을 것 같은데요. 테스트는 프로덕션 코드가 의도한대로 잘 동작하는지 확인하기 위함입니다. 현재 CarsGeneratorTest에서는 Car를 생성 후 이름이 같은 또 다른 Car 객체를 생성하여 동일한 객체인지 비교하고 있습니다. 이러한 Car 비교가 도메인 로직에서 사용되는 곳이 없고 테스트에서만 사용되었는데요. 그렇다면 두 가지 고민을 해볼 것 같습니다.

1. Car 객체를 비교하는 테스트가 필요한 이유는 무엇인가?
2. Car 객체를 비교하는 기준은 무엇인가

1번: 프로덕션 코드에서는 안 쓰이지만 이런 테스트를 만들고 싶다. -> 그 이유는?
2번: Car의 이름만 같으면 같은 Car로 인식, name, position이 모두 같아야 같은 Car로 인식 등
- 이름만 같아도 된다면 getName으로 비교해도 가능 -> 이것이 정말 원하는 비교방식 맞는지

이러한 질문을 해보시고 답을 찾아가다 보면 조금 정리가 되지 않을까 싶습니다 :)

sonypark commented on src/main/java/racingcar/service/RacingService.java:
```diff
@@ -0,0 +1,52 @@
+		for (int i = 0; i < round; i++) {
+			raceRound();
+		}
+	}
```
👍

sonypark commented on src/main/java/racingcar/service/RacingService.java:
```diff
@@ -0,0 +1,52 @@
+			.mapToInt(Car::getPosition)
+			.max()
+			.orElseThrow(IllegalStateException::new);
+	}
```
IllegalStateException 를 throw 해주신 이유가 있나요?
IllegalStateException이란 언제 발생할까요?🤔

yxxnghwan replied:
IllegalStateException은 적절하지 않은 것 같네요! 소니 덕분에 자바의 예외들이 어떨 때 발생하는지 짚어보게 되었어요.

해당 부분은 Car 클래스의 생성자에서 position을 초기화 하기 때문에 빈 옵셔널이 반환될 일이 없을 것 같아요.
그래도 혹시 캐치하지 못한 상황이 있을 수도 있으니 get() 보단 orElse(MIN_POSITION) 으로 변경했습니다.

sonypark commented on src/main/java/racingcar/ui/RacingCarInput.java:
```diff
@@ -0,0 +1,51 @@
+	private static final String CAR_NAME_INPUT_MESSAGE
+		= "경주할 자동차 이름을 입력하세요(이름은 쉼표(,)를 기준으로 구분).";
+	private static final String ROUND_INPUT_MESSAGE
+		= "시도할 회수는 몇회인가요?";
```
개행은 의도하신건가요?ㅎㅎ

yxxnghwan replied:
해당 부분은 개행할 필요가 없어보이네요! 지적 감사합니다.

sonypark commented on src/main/java/racingcar/validator/CarNameValidator.java:
```diff
@@ -0,0 +1,20 @@
+		if (isIllegalLength(name)) {
+			throw new IllegalArgumentException(MAX_NAME_LENGTH_ERROR_MESSAGE);
+		}
+	}
```
알렉스가 말씀하신대로 name.isBlank()를 메서드로 뺄 이유는 없습니다. 이미 String에서 제공하는 메서드를 사용하고 있어 오히려 새로운 메서드로 빼는것은 불필요한 메서드를 생성하는 것이라 생각합니다. 잘 하셨어요! 👍

sonypark commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,72 @@
+		if (isNullOrEmpty(input)) {
+			return 0;
+		}
+		List<String> numberStringList = split(input);
```
split 메서드는 입력값에 대한 검증과 분리 로직을 수행하는 메서드네요.
따라서 목적에 맞게 CalculatorInput에서 수행하는건 어떨까요?

yxxnghwan replied:
해당 피드백을 보고나서 한 클래스에서 너무 많은 역할을 맡고 있다는 걸 알았어요.
StringCalculator는 계산 기능만 하도록 변경했고, 검증과 분리하는 클래스를 구현해서 역할을 나눠주었습니다.

거기에 따라 StringCalculator.calculate() 메서드의 시그니처도 변경됐고, 그에 따른 테스트코드도 다시 작성했어요!

sonypark reviewed (commented):
알렉스 안녕하세요! 이번 미션 리뷰어를 맡게된 소니입니다!😊
커밋로그도 꼼꼼하게 잘 작성해주셨고, 테스트 코드도 잘 작성해주셨습니다.
TDD기반으로 작성하시려고 노력한 모습이 보이네요👍
메서드 분리도 목적에 맞게 잘 해주셨어요! 💯
몇가지 코멘트 남겼으니 확인 부탁드릴게요 :)

sonypark reviewed (changes requested):
comment로는 슬랙 알림이 가지 않는 것 같아 Request changes로 한번 더 요청드립니다 :)

yxxnghwan committed:
docs(racing-car): 구현할 기능 목록 수정 (체크박스, 세부적으로 나눔)

구현할 기능 목록에 요소마다 체크 박스를 두고 구현시 체크하도록 변경
한 요소에 여러 기능이 있는 경우 요소를 나눔

yxxnghwan committed:
docs(calculate): 구현할 기능 목록 수정 (체크박스, 세부적으로 나눔)

구현할 기능 목록에 요소마다 체크 박스를 두고 구현시 체크하도록 변경
한 요소에 여러 기능이 있는 경우 요소를 나눔

yxxnghwan committed:
refactor(racing-car): StringCalculator.toNumberList 불필요한 박싱, 언박싱 제거

yxxnghwan committed:
refactor(racing-car): static 키워드 누락되었던 것 추가

yxxnghwan committed:
refactor(racing-car): RacingCarInput 상수선언 개행 변경

yxxnghwan committed:
refactor(racing-car): 옵셔널 처리 변경

yxxnghwan committed:
test(calculator): StringCalculator.calculate 시그니처 변경에 의한 테스트 코드 작성

기존 String 타입을 인자로 받아서 검증, 분리, 계산을 진행
역할을 나눠서 분리된 값을 받아 계산만 진행하도록 클래스 설계를 변경
그에 따른 새로운 테스트 코드 작성

yxxnghwan committed:
test(calculator): StringNumberValidator에 대한 테스트 코드 작성

yxxnghwan committed:
refactor(calculator): StringCalculator에서 검증 로직을 StringNumberValidator로 분리

yxxnghwan committed:
test(calculator): UserInputValidator가 사용자의 입력값이 커스텀 구분자인지 아닌지 판단하는 부분 테스트 코드 작성

yxxnghwan committed:
refactor(calculator): StringCalculator에서 사용자의 입력을 검증하는 기능을 UserInputValidator로 이동

yxxnghwan committed:
test(calculator): UserInputValidator가 null과 whitespace를 잘 걸러내는지 테스트 작성

yxxnghwan committed:
test(calculator): StringCalculator가 수행하던 분리 로직을 수행할 FormulaStringSplitter의 분리 기능 테스트

yxxnghwan committed:
feat(calculator): FormulaStringSplitter 구현

yxxnghwan committed:
feat(calculator): 사용자 입력 형태를 클래스로 정의

yxxnghwan committed:
feat(calculator): 변경된 사항에 따라 CalculatorInput 기능 다시 구현

UserInputValidator의 메서드를 호출하여 검증하고 적절한 UserInputForm을 반환하도록 변경

yxxnghwan committed:
feat(calculator): 변경된 사항에 따라 메인 로직 변경, 콘솔창 입력이 의도대로 되지 않던 오류 수정

yxxnghwan committed:
chore(calculator): 패키지 정리 및 불필요한 import문 삭제

yxxnghwan committed:
style(all): 우테코 스타일 컨벤션 적용

yxxnghwan commented:
소니가 준 피드백 내용 반영해서 미션 수정해봤어요!
추가적으로 손봐야할 부분이 있다면 알려주세요!

yxxnghwan committed:
refactor(racing-car): 테스트 코드 리팩토링

테스트 메서드 이름을 영어로 붙이고 @DisplayName을 활용해서 warning이 발생하지 않도록 수정

하나의 테스트 메서드에서 여러 케이스를 테스트하는 코드들 각각 테스트 메서드로 분리

yxxnghwan committed:
refactor(racing-car): RacingServiceTest 테스트케이스 준비 과정 분리

@BeforeEach, @AfterEach 활용해서 각 테스트 케이스 시행마다 경주에 참여할 자동차들 세팅

sonypark commented on src/main/java/racingcar/service/RacingService.java:
```diff
@@ -0,0 +1,52 @@
+		RacingCarOutput.printRoundResult(cars);
+		System.out.println();
+	}
+
```
1. racdRound 메서드는 어떻게 테스트 할 수 있을까요?
2. 도메인 로직과 View 로직이 혼합되어 있는데 분리해볼 수 있을까요?

sonypark replied:
이 부분은 2단계 미션에서 구현하셔도 됩니다!

sonypark commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,48 @@
+		if (randomNumber >= PROCEED_FlAG_NUMBER) {
+			this.position++;
+		}
+	}
```
파라미터명에 random이 들어갈 필요가 있을까요?

sonypark commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,48 @@
+			this.position++;
+		}
+	}
+
+	public boolean isWinner(int maxPosition) {
+		return this.position == maxPosition;
+	}
+
```
이 메서드명이 적절할까요?

sonypark commented on src/main/java/calculator/CalculatorMain.java:
```diff
@@ -0,0 +1,14 @@
+
+public class CalculatorMain {
+	public static Scanner sc = new Scanner(System.in);
+
```
Scanner가 여기에 있는게 맞을까요?

sonypark reviewed (changes requested):
알렉스 안녕하세요~
피드백 잘 반영해주신 것 확인했습니다. 수고하셨어요!👍
여기서 조금 더 개선해봅시다. 💪
2차 피드백 드렸으니 확인해 부탁드려요~
질문은 댓글에 남겨주시면 답변 드리겠습니다 🙏🏻

sonypark commented:
피드백 남긴 부분은 2단계에서 반영해주셔도 될 것 같아서 1단계는 머지하도록 하겠습니다.
2단계도 화이팅! 💪

yxxnghwan commented:
감사합니다 소니!😃
덕분에 많이 공부했어요! 2단계도 열심히 하겠습니다!💪

저 하나만 여쭤볼 수 있을까요?

0 이상의 정수를 판단하는 로직에서 정규표현식을 사용하지 않고

```java
try {
   Integer.parseInt(number);
} catch(NumberFormatException e) {
    return false;
}

if (number < 0) {
    return false;
}

return true;
```
이렇게 예외를 캐치해서 판단하게 로직을 짜면 문제가 있을까요?

sonypark commented:
> 감사합니다 소니!😃 덕분에 많이 공부했어요! 2단계도 열심히 하겠습니다!💪
> 
> 저 하나만 여쭤볼 수 있을까요?
> 
> 0 이상의 정수를 판단하는 로직에서 정규표현식을 사용하지 않고
> 
> ```java
> try {
>    Integer.parseInt(number);
> } catch(NumberFormatException e) {
>     return false;
> }
> 
> if (number < 0) {
>     return false;
> }
> 
> return true;
> ```
> 
> 이렇게 예외를 캐치해서 판단하게 로직을 짜면 문제가 있을까요?

이렇게 해도 기대한대로 동작할 것 같네요. 사실상 코드는 의도한대로 동작하는게 중요하죠 :)
다만 개발자는 조금 더 객체지향적이고, 유지보수성이 높고, 가독성 좋은 코드를 짜기 위해 고민해야하는거구요.
일단 위 로직은 두 가지 일을 하고 있네요. 숫자인지 판단하고, 숫자가 아니면 false를 리턴하는군요(숫자가 아니라서 false). 아래 로직은 0보다 작은 값이면 false를 리턴하네요. 이 메서드의 이름은 무엇인가요? 숫자를 받아서 0보다 크냐 작냐만 판단해서 리턴해줄 수도 있지 않을까요?

yxxnghwan commented:
> > 감사합니다 소니!😃 덕분에 많이 공부했어요! 2단계도 열심히 하겠습니다!💪
> > 저 하나만 여쭤볼 수 있을까요?
> > 0 이상의 정수를 판단하는 로직에서 정규표현식을 사용하지 않고
> > ```java
> > try {
> >    Integer.parseInt(number);
> > } catch(NumberFormatException e) {
> >     return false;
> > }
> > 
> > if (number < 0) {
> >     return false;
> > }
> > 
> > return true;
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 이렇게 예외를 캐치해서 판단하게 로직을 짜면 문제가 있을까요?
> 
> 이렇게 해도 기대한대로 동작할 것 같네요. 사실상 코드는 의도한대로 동작하는게 중요하죠 :) 다만 개발자는 조금 더 객체지향적이고, 유지보수성이 높고, 가독성 좋은 코드를 짜기 위해 고민해야하는거구요. 일단 위 로직은 두 가지 일을 하고 있네요. 숫자인지 판단하고, 숫자가 아니면 false를 리턴하는군요(숫자가 아니라서 false). 아래 로직은 0보다 작은 값이면 false를 리턴하네요. 이 메서드의 이름은 무엇인가요? 숫자를 받아서 0보다 크냐 작냐만 판단해서 리턴해줄 수도 있지 않을까요?

아! 문자열을 판단하는 로직이었는데 중간에 파싱을 빠뜨렸네요!
소니가 말한대로 저렇게 되면 한 메서드에서 두 가지를 판단하게 될 것 같아요.
한번 위에 검증과 아래 검증 분리해봤어요.
```java
public boolean isRightNumber(String number) {
    if (isNumeric(number)) {
        return false;
    }

    return Integer.parseInt(number) >= 0;
}

private boolean isNumeric(String number) {
    try {
        Integer.parseInt(number);
    } catch (NumberFormatException e) {
        return true;
    }
    return false;
}
```

sonypark commented:
Integer.parseInt(number) 가 중복으로 쓰이고 있네요. isRightNumber()의 결과값이 또  Integer.parseInt(number)를 사용하지는 않나요?
위 메서드를 사용하면 음수나 문자가 들어왔을 때 예외는 발생하지 않을 것 같은데 맞을까요?

yxxnghwan commented:
> Integer.parseInt(number) 가 중복으로 쓰이고 있네요. isRightNumber()의 결과값이 또 Integer.parseInt(number)를 사용하지는 않나요? 위 메서드를 사용하면 음수나 문자가 들어왔을 때 예외는 발생하지 않을 것 같은데 맞을까요?

아 중복이군요... 전 바보인가봐요... 결과값에도 결국 Integer.parseInt()를 사용합니다.
결과를 파싱하는 것과 예외발생까지 합쳐서 아래처럼 쓸 수 있을 것 같아요.
```java
public int parseNumber(String number) {
    try {
        return getNaturalNumber(number);
    } catch (NumberFormatException e) {
        throw new IllegalArgumentException("숫자가 아님.");
    }
}

private int getNaturalNumber(String number) {
    int n = Integer.parseInt(number);
    if (n > 0) {
        return n;
    }
    throw new IllegalArgumentException("숫자이지만 자연수 아님.");
}
```
이 부분은 좀 더 고민을 해봐야 할 것 같아요🤔
try-catch를 사용하면, '정수가 아니면 거른다' 하는 생각을 바로 코드로 옮길 수 있어서 단순화될줄 알았는데,
오히려 꼬여버리고 catch때문에 분기문에 제약이 걸려서 메서드의 역할을 나누기도 어려워지네요...
일단 인덴트를 한칸 먹고 시작하니 그것도 거슬리고 ㅠㅠ
굳이 필요없는 try문은 피해야할 것 같습니다...

같이 쓸데없는거 고민해줘서 고마워요 소니 ㅠㅠ😭

sonypark commented:
네 이런 고민을 하는것 자체가 더 좋은 코드로 나아가는 첫걸음이라고 생각합니다. 미션을 진행함에 있어서 쓸데 없는 고민은 없습니다! 어떤 질문이든 좋아요!(미션과 너무 관련이 없는 질문만 아니라면?!) 하나의 메서드는 하나의 역할을 하도록 나눠보길 권장 드립니다 :) 앞으로 미션 진행하며 이런 저런 질문이 많이 생길텐데 마음껏 질문 주세요~
